File: harness/decision_capsule_contract/closure.py
"""Outer StructuralReplayClosure sealing and validation."""

from __future__ import annotations

import hashlib

from kernel_operational_contract.shape import bare_hash

from .branch import (_derive_with_validated_capsule, _local_preflight as
                     _validate_branch_local,
                     _validate_comparison, _validate_with_validated_capsule)
from .capsule import (_local_preflight as _validate_capsule_local,
                      validate_decision_capsule)
from .codec import (ContractError, bounded_deepcopy, canonical_json,
                    decode_canonical_json, exact_object_shell,
                    require_constants, require_identity_strings,
                    validate_json_tree)
from .constants import *  # noqa: F403 - frozen vocabulary is intentionally central
from .shape import attestations, reflection_refs


def _identity(closure: dict[str, object], allow_blank: bool) -> None:
    if allow_blank and closure["closure_id"] == closure["closure_sha256"] == "":
        return
    digest = bare_hash(closure["closure_sha256"], "closure_sha256")
    if closure["closure_id"] != f"{CLOSURE_PREFIX}{digest}":  # noqa: F405
        raise ContractError("closure_id must bind closure_sha256")


def _local_preflight(value: object, *, allow_blank: bool = False,
                     ignore_identity: bool = False) -> dict[str, object]:
    closure = exact_object_shell(  # noqa: F405
        value, CLOSURE_FIELDS, "StructuralReplayClosure")
    constants = {"api_version": CLOSURE_API, "canonicalization": CANONICALIZATION,  # noqa: F405
                 "kind": CLOSURE_KIND, "result": SUCCESS_MARKER}  # noqa: F405
    require_constants(closure, constants)
    require_identity_strings(closure, "closure_id", "closure_sha256")
    if not ignore_identity:
        _identity(closure, allow_blank)
    validate_json_tree(closure["attestations"], MAX_CLOSURE_BYTES)  # noqa: F405
    attestations(closure["attestations"])
    validate_json_tree(  # noqa: F405
        closure["reflection_report_artifact_refs"], MAX_CLOSURE_BYTES)
    reflection_refs(closure["reflection_report_artifact_refs"])
    _validate_branch_local(closure["evaluation_branch"])
    _validate_capsule_local(closure["decision_capsule"])
    validate_json_tree(closure, MAX_CLOSURE_BYTES)  # noqa: F405
    return closure


def _shape(value: object, *, allow_blank: bool = False,
           ignore_identity: bool = False) -> dict[str, object]:
    closure = _local_preflight(
        value, allow_blank=allow_blank, ignore_identity=ignore_identity)
    _validate_comparison(
        closure["evaluation_branch"], closure["decision_capsule"])
    capsule = validate_decision_capsule(closure["decision_capsule"])
    _validate_with_validated_capsule(closure["evaluation_branch"], capsule)
    return closure


def _blanked(value: object) -> dict[str, object]:
    exact_object_shell(value, CLOSURE_FIELDS, "closure digest input")  # noqa: F405
    validate_json_tree(value, MAX_CLOSURE_BYTES)  # noqa: F405
    result = bounded_deepcopy(value, MAX_CLOSURE_BYTES)  # noqa: F405
    result["closure_id"], result["closure_sha256"] = "", ""
    return result


def structural_replay_closure_digest(value: object) -> str:
    closure = _shape(value, ignore_identity=True)
    blank = _blanked(closure)
    raw = canonical_json(blank)
    if len(raw) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure blank preimage exceeds its byte ceiling")
    return hashlib.sha256(CLOSURE_DOMAIN + raw).hexdigest()  # noqa: F405


def validate_structural_replay_closure(value: object) -> dict[str, object]:
    closure = _shape(value)
    blank = _blanked(closure)
    raw = canonical_json(blank)
    if len(canonical_json(closure)) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure exceeds its byte ceiling")
    digest = hashlib.sha256(CLOSURE_DOMAIN + raw).hexdigest()  # noqa: F405
    if closure["closure_sha256"] != digest:
        raise ContractError("closure_sha256 does not match canonical preimage")
    return closure


def seal_structural_replay_closure(value: object) -> dict[str, object]:
    if (type(value) is not dict or value.get("closure_id") != "" or
            value.get("closure_sha256") != ""):
        raise ContractError("sealing outer closure requires blank own identity")
    shaped = _shape(value, allow_blank=True)
    closure = bounded_deepcopy(shaped, MAX_CLOSURE_BYTES)  # noqa: F405
    raw = canonical_json(closure)
    if len(raw) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure blank preimage exceeds its byte ceiling")
    digest = hashlib.sha256(CLOSURE_DOMAIN + raw).hexdigest()  # noqa: F405
    closure["closure_id"] = f"{CLOSURE_PREFIX}{digest}"  # noqa: F405
    closure["closure_sha256"] = digest
    if len(canonical_json(closure)) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure exceeds its sealed byte ceiling")
    _identity(closure, False)
    return closure


def derive_structural_replay_closure(
        capsule: object, reflection_report_artifact_refs: object) -> dict[str, object]:
    validate_json_tree(
        reflection_report_artifact_refs, MAX_CLOSURE_BYTES)  # noqa: F405
    refs = reflection_refs(reflection_report_artifact_refs)
    _validate_capsule_local(capsule)
    validated = validate_decision_capsule(capsule)
    candidate = {
        "api_version": CLOSURE_API, "attestations": {  # noqa: F405
            field: False for field in ATTESTATION_FIELDS},  # noqa: F405
        "canonicalization": CANONICALIZATION, "closure_id": "",  # noqa: F405
        "closure_sha256": "", "decision_capsule": bounded_deepcopy(
            validated, MAX_CAPSULE_BYTES),  # noqa: F405
        "evaluation_branch": _derive_with_validated_capsule(validated),
        "kind": CLOSURE_KIND,  # noqa: F405
        "reflection_report_artifact_refs": bounded_deepcopy(
            refs, MAX_CLOSURE_BYTES),  # noqa: F405
        "result": SUCCESS_MARKER,  # noqa: F405
    }
    raw = canonical_json(candidate)
    if len(raw) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure blank preimage exceeds its byte ceiling")
    digest = hashlib.sha256(CLOSURE_DOMAIN + raw).hexdigest()  # noqa: F405
    candidate["closure_id"] = f"{CLOSURE_PREFIX}{digest}"  # noqa: F405
    candidate["closure_sha256"] = digest
    if len(canonical_json(candidate)) > MAX_CLOSURE_BYTES:  # noqa: F405
        raise ContractError("outer closure exceeds its sealed byte ceiling")
    _identity(candidate, False)
    return candidate


def decode_structural_replay_closure(raw: bytes) -> dict[str, object]:
    value = decode_canonical_json(raw, MAX_CLOSURE_BYTES)  # noqa: F405
    return validate_structural_replay_closure(value)

File: forge-core/internal/decisioncapsulecontract/closure.go
package decisioncapsulecontract

import (
	"fmt"

	op "forgeos/forge-core/internal/kerneloperationalcontract"
)

func validateReflectionRefs(values []op.ArtifactRef) error {
	if values == nil || len(values) > maxReflectionReportRefs {
		return fmt.Errorf("reflection_report_artifact_refs cardinality must be 0..%d",
			maxReflectionReportRefs)
	}
	keys := make([]string, len(values))
	for index := range values {
		if err := validateArtifactRef(&values[index]); err != nil {
			return fmt.Errorf("reflection_report_artifact_refs[%d]: %w", index, err)
		}
		if values[index].ArtifactKind != "reflection_report" {
			return fmt.Errorf("reflection report ArtifactRefs require reflection_report kind")
		}
		raw, _ := op.CanonicalJSON(&values[index])
		keys[index] = string(raw)
	}
	return validateUniqueKeys(keys, "reflection_report_artifact_refs", true)
}

func validateClosureLocal(value *StructuralReplayClosure, allowBlank bool,
	ignoreIdentity bool) error {
	if value == nil {
		return fmt.Errorf("StructuralReplayClosure is nil")
	}
	if value.APIVersion != closureAPI || value.Canonicalization != canonicalization ||
		value.Kind != closureKind || value.Result != successMarker {
		return fmt.Errorf("StructuralReplayClosure constants or result marker differ")
	}
	if !ignoreIdentity {
		if err := validateIdentity(value.ClosureID, value.ClosureSHA256,
			closurePrefix, "outer closure", allowBlank); err != nil {
			return err
		}
	}
	if err := validateReplayAttestations(value.Attestations); err != nil {
		return err
	}
	if err := validateReflectionRefs(value.ReflectionReportArtifactRefs); err != nil {
		return err
	}
	if err := validateBranchLocal(&value.EvaluationBranch, false); err != nil {
		return fmt.Errorf("evaluation_branch: %w", err)
	}
	if err := validateCapsuleLocal(&value.DecisionCapsule, false, false); err != nil {
		return fmt.Errorf("decision_capsule: %w", err)
	}
	return nil
}

func validateClosureShape(value *StructuralReplayClosure, allowBlank bool) error {
	if err := validateClosureLocal(value, allowBlank, false); err != nil {
		return err
	}
	if err := requireBranchComparison(
		&value.EvaluationBranch, &value.DecisionCapsule); err != nil {
		return fmt.Errorf("evaluation_branch: %w", err)
	}
	if _, err := measureTypedJSON(value, maxClosureBytes); err != nil {
		return err
	}
	if _, err := validateCapsulePrepared(&value.DecisionCapsule, false); err != nil {
		return fmt.Errorf("decision_capsule: %w", err)
	}
	if err := validateBranchPrepared(&value.EvaluationBranch,
		&value.DecisionCapsule, false); err != nil {
		return fmt.Errorf("evaluation_branch: %w", err)
	}
	return nil
}

func deriveClosurePrepared(capsule *DecisionCapsule,
	reflectionReportArtifactRefs []op.ArtifactRef) (*StructuralReplayClosure, error) {
	branch, err := deriveBranchPrepared(capsule)
	if err != nil {
		return nil, err
	}
	closure := &StructuralReplayClosure{
		APIVersion:                   closureAPI,
		Canonicalization:             canonicalization,
		DecisionCapsule:              *capsule,
		EvaluationBranch:             *branch,
		Kind:                         closureKind,
		ReflectionReportArtifactRefs: append([]op.ArtifactRef{}, reflectionReportArtifactRefs...),
		Result:                       successMarker,
	}
	if err := validateReflectionRefs(closure.ReflectionReportArtifactRefs); err != nil {
		return nil, err
	}
	digest, err := digestValue(closure, closureDomain, maxClosureBytes)
	if err != nil {
		return nil, fmt.Errorf("outer closure blank preimage: %w", err)
	}
	closure.ClosureID, closure.ClosureSHA256 = closurePrefix+digest, digest
	if _, err := canonicalBytes(closure, maxClosureBytes); err != nil {
		return nil, fmt.Errorf("sealed outer closure: %w", err)
	}
	return closure, nil
}

func StructuralReplayClosureDigest(value *StructuralReplayClosure) (string, error) {
	if value == nil {
		return "", fmt.Errorf("StructuralReplayClosure is nil")
	}
	if err := validateClosureLocal(value, false, true); err != nil {
		return "", err
	}
	if _, err := measureTypedJSON(value, maxClosureBytes); err != nil {
		return "", err
	}
	blank := *value
	blank.ClosureID, blank.ClosureSHA256 = "", ""
	if err := validateClosureShape(&blank, true); err != nil {
		return "", err
	}
	return digestValue(&blank, closureDomain, maxClosureBytes)
}

func ValidateStructuralReplayClosure(value *StructuralReplayClosure) error {
	if err := validateClosureShape(value, false); err != nil {
		return err
	}
	blank := *value
	blank.ClosureID, blank.ClosureSHA256 = "", ""
	digest, err := digestValue(&blank, closureDomain, maxClosureBytes)
	if err != nil {
		return fmt.Errorf("outer closure blank preimage: %w", err)
	}
	if value.ClosureSHA256 != digest {
		return fmt.Errorf("closure_sha256 does not match canonical preimage")
	}
	_, err = canonicalBytes(value, maxClosureBytes)
	return err
}

func SealStructuralReplayClosure(value *StructuralReplayClosure) (
	*StructuralReplayClosure, error,
) {
	if value == nil || value.ClosureID != "" || value.ClosureSHA256 != "" {
		return nil, fmt.Errorf("sealing outer closure requires blank own identity")
	}
	if err := validateClosureLocal(value, true, false); err != nil {
		return nil, err
	}
	if err := validateClosureShape(value, true); err != nil {
		return nil, err
	}
	sealed, err := cloneValue(value, maxClosureBytes)
	if err != nil {
		return nil, err
	}
	digest, err := digestValue(sealed, closureDomain, maxClosureBytes)
	if err != nil {
		return nil, fmt.Errorf("outer closure blank preimage: %w", err)
	}
	sealed.ClosureID, sealed.ClosureSHA256 = closurePrefix+digest, digest
	if _, err := canonicalBytes(sealed, maxClosureBytes); err != nil {
		return nil, fmt.Errorf("sealed outer closure: %w", err)
	}
	return sealed, nil
}

func DeriveStructuralReplayClosure(capsule *DecisionCapsule,
	reflectionReportArtifactRefs []op.ArtifactRef) (*StructuralReplayClosure, error) {
	if err := validateReflectionRefs(reflectionReportArtifactRefs); err != nil {
		return nil, err
	}
	if err := validateCapsuleLocal(capsule, false, false); err != nil {
		return nil, err
	}
	if _, err := validateCapsulePrepared(capsule, false); err != nil {
		return nil, err
	}
	cloned, err := cloneValue(capsule, maxCapsuleBytes)
	if err != nil {
		return nil, err
	}
	return deriveClosurePrepared(cloned, reflectionReportArtifactRefs)
}

func DecodeStructuralReplayClosure(raw []byte) (*StructuralReplayClosure, error) {
	value, err := decodeExact[StructuralReplayClosure](raw, maxClosureBytes)
	if err != nil {
		return nil, err
	}
	return value, ValidateStructuralReplayClosure(value)
}
